### PR TITLE
Docs + Examples + Tests 

### PR DESCRIPTION
**Pull request by arthurbr11**

### CODE REVIEW

File: sentence_transformers/losses/DistillKLDivLoss.py
Comment (tomaarsen) [inline sentence_transformers/losses/DistillKLDivLoss.py:1]: This file is not in `losses.md` for ST

File: sentence_transformers/losses/MarginMSELoss.py [hunk 1]
```diff
@@ -111,7 +113,7 @@ def compute_labels(batch):
                 )
                 trainer.train()
 
-            For multiple negatives:
+            We  can also use multiple negatives during the knowledge distillation.
```
Comment (tomaarsen) [inline sentence_transformers/losses/MarginMSELoss.py:116]:
```suggestion
            We can also use multiple negatives during the knowledge distillation.
```

File: sentence_transformers/losses/MarginMSELoss.py [hunk 2]
```diff
@@ -120,14 +122,20 @@ def compute_labels(batch):
...
+                        "query": ["It's nice weather outside today.", "He drove to work."],
+                        "passage1": ["It's so sunny.", "He took the car to work."],
+                        "passage2": ["It's very cold.", "She walked to the store."],
+                        "passage3": ["Its rainy", "She took the bus"],
+                    }
+                )
+
 
                 def compute_labels(batch):
+                    import torch
```
Comment (tomaarsen) [inline sentence_transformers/losses/MarginMSELoss.py:137]: Maybe move this outside of the function to the top? If we even need it

File: sentence_transformers/sparse_encoder/SparseEncoder.py [hunk 3]
```diff
@@ -75,6 +75,109 @@ class SparseEncoder(SentenceTransformer):
 
     # -----------------------------------Added------------------------------------
     # - get_sparsity_stats
+    """
+    Loads or creates a SparseEncoder model that can be used to map sentences / text to sparse embeddings.
+
+    Args:
+        model_name_or_path (str, optional): If it is a filepath on disc, it loads the model from that path. If it is not a path,
+            it first tries to download a pre-trained SentenceTransformer model. If that fails, tries to construct a model
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/SparseEncoder.py:83]:
```suggestion
            it first tries to download a pre-trained SparseEncoder model. If that fails, tries to construct a model
```
It's smart to ctrl+F "SentenceTransformer" on the docs page and to try and resolve those

File: sentence_transformers/sparse_encoder/SparseEncoder.py [hunk 4]
```diff
@@ -134,6 +237,55 @@ def encode(
         device: str | None = None,
         **kwargs: Any,
     ) -> list[Tensor] | np.ndarray | Tensor | dict[str, Tensor] | list[dict[str, Tensor]]:
+        """
+        Computes sentence embeddings.
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/SparseEncoder.py:241]:
```suggestion
        Computes sparse sentence embeddings.
```
Might be a bit clearer?

File: sentence_transformers/sparse_encoder/SparseEncoder.py [hunk 5]
```diff
@@ -134,6 +237,55 @@ def encode(
...
+                sentence "What is the capital of France?" will be encoded as "query: What is the capital of France?"
+                because the sentence is appended to the prompt. If ``prompt`` is set, ``prompt_name`` is ignored. Defaults to None.
+            batch_size (int, optional): The batch size used for the computation. Defaults to 32.
+            show_progress_bar (bool, optional): Whether to output a progress bar when encode sentences. Defaults to None.
+            output_value (Optional[Literal["sentence_embedding", "token_embeddings"]], optional): The type of embeddings to return:
+                "sentence_embedding" to get sentence embeddings, "token_embeddings" to get wordpiece token embeddings, and `None`,
+                to get all output values. Defaults to "sentence_embedding".
+            convert_to_tensor (bool, optional): Whether the output should be one large tensor. Overwrites `convert_to_numpy`.
+                Defaults to False.
+            convert_to_sparse_tensor (bool, optional): Whether the output should be in the format of a sparse tensor
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/SparseEncoder.py:260]:
```suggestion
            convert_to_sparse_tensor (bool, optional): Whether the output should be in the format of a sparse tensor.
```
Do we want to mention which sparse format we're using?

File: sentence_transformers/sparse_encoder/SparseEncoder.py [hunk 6]
```diff
@@ -134,6 +237,55 @@ def encode(
...
+                Defaults to True.
+            save_on_cpu (bool, optional):  Whether the output should be move on cpu or stay on the device it have been compute on.
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/SparseEncoder.py:262]:
```suggestion
            save_to_cpu (bool, optional):  Whether the output should be moved to cpu or stay on the device it has been computed on.
```
I think `save_to_cpu` is clearer. Or perhaps `move_to_cpu`?

File: sentence_transformers/sparse_encoder/SparseEncoder.py [hunk 7]
```diff
@@ -134,6 +237,55 @@ def encode(
...
+                Defaults to False
+            device (str, optional): Which :class:`torch.device` to use for the computation. Defaults to None.
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/SparseEncoder.py:264]: Maybe worth explaining that the current device gets used if None.

File: sentence_transformers/sparse_encoder/callbacks/splade_callbacks.py [hunk 8]
```diff
@@ -25,6 +25,11 @@ class SpladeLambdaSchedulerCallback(TrainerCallback):
 
     The scheduler gradually increases the lambda values from 0 to their max value
     within the specified warmup ratio of the total training steps.
+
+     Args:
+            loss: SpladeLoss instance to be updated
+            scheduler_type: Type of scheduler ('linear' or 'quadratic')
+            warmup_ratio: Ratio of total steps to reach max lambda values (default: 1/3)
     """
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/callbacks/splade_callbacks.py:33]: I think the entire docstring here can move to the `__init__`, and then perhaps we can add a `callback.md` akin to e.g. `sampler.md` that just shows this one Callback?
It might be useful to inspire people that callbacks can be used for stuff like this, and not just e.g. logging

File: docs/quickstart.rst [hunk 9]
```diff
@@ -151,6 +151,67 @@ The usage for Cross Encoder (a.k.a. reranker) models is similar to Sentence Tran
...
+   print(embeddings.shape)
+   # [3, 30522] - sparse representation with vocabulary size dimensions
+
+   # 3. Calculate the embedding similarities (using dot product by default)
+   similarities = model.similarity(embeddings, embeddings)
+   print(similarities)
+   # tensor([[3.5629e+01, 9.1541e+00, 1.1269e-01],
+   #         [9.1541e+00, 2.7478e+01, 1.9061e-02],
+   #         [1.1269e-01, 1.9061e-02, 2.9612e+01]], device='cuda:0')
+
```
Comment (tomaarsen) [inline docs/quickstart.rst:206]:
```suggestion
   print(similarities)
   # tensor([[   35.6293,     9.1541,     0.1127],
   #         [    9.1541,    27.4778,     0.0191],
   #         [    0.1127,     0.0191,    29.6122]], device='cuda:0')
```
I just figured out that I can use:
```python
import torch
torch.set_printoptions(sci_mode=False)  # Disable scientific notation for better readability
```
and then our examples are a bit clearer. We shouldn't have the `torch.set_printoptions` anywhere visible, but the more readable outputs are quite useful.

File: docs/quickstart.rst [hunk 10]
```diff
@@ -151,6 +151,67 @@ The usage for Cross Encoder (a.k.a. reranker) models is similar to Sentence Tran
...
+
+   # 4. Check sparsity statistics
+   stats = SparseEncoder.get_sparsity_stats(embeddings)
+   print(f"Sparsity: {stats['row_sparsity_mean']:.2%}")  # Typically >99% zeros
+   print(f"Avg non-zeros per embedding: {stats['row_non_zero_mean']}")
```
Comment (tomaarsen) [inline docs/quickstart.rst:211]:
```suggestion
   print(f"Avg non-zeros per embedding: {stats['row_non_zero_mean']:.2f}")
```

File: sentence_transformers/sparse_encoder/SparseEncoder.py [hunk 11]
```diff
@@ -374,7 +526,9 @@ def _load_auto_model(
         Returns:
             List[nn.Module]: A list containing the transformer model and the pooling model.
         """
-        logger.warning(f"No sparse-encoder model found with name {model_name_or_path}. Creating a new one.")
+        logger.warning(
+            f"No sparse-encoder model found with name {model_name_or_path}. Creating a new one with defaults settings compatible to the based model."
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/SparseEncoder.py:530]:
```suggestion
            f"No sparse-encoder model found with name {model_name_or_path}. Creating a new one with defaults settings compatible to the base model."
```
haha, kind of a funny typo

File: docs/package_reference/sparse_encoder/evaluation.md [hunk 12]
```diff
@@ -0,0 +1,51 @@
+# Evaluation
+
+`sentence_transformers.sparse_encoder.evaluation` defines different classes, that can be used to evaluate the SparseEncoder model during training.
+
+## SparseInformationRetrievalEvaluator
+```{eval-rst}
+.. autoclass:: sentence_transformers.sparse_encoder.evaluation.SparseInformationRetrievalEvaluator
+   :members:
```
[the remark this answers is not in the document]
Comment (arthurbr11) [reply docs/package_reference/sparse_encoder/evaluation.md:8]: Yes i just get rid of :members: as it's done in the ST doc for evaluation

File: docs/package_reference/sparse_encoder/evaluation.md [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+```
+
+## SparseNanoBEIREvaluator
+```{eval-rst}
+.. autoclass:: sentence_transformers.sparse_encoder.evaluation.SparseNanoBEIREvaluator
+   :members:
```
[the remark this answers is not in the document]
Comment (arthurbr11) [reply docs/package_reference/sparse_encoder/evaluation.md:14]: patch with what's say above

File: docs/package_reference/sparse_encoder/evaluation.md [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+## SparseTripletEvaluator
+```{eval-rst}
+.. autoclass:: sentence_transformers.sparse_encoder.evaluation.SparseTripletEvaluator
+   :members:
+```
+
+## SparseRerankingEvaluator
+```{eval-rst}
+.. autoclass:: sentence_transformers.sparse_encoder.evaluation.SparseRerankingEvaluator
+   :members:
```
[the remark this answers is not in the document]
Comment (arthurbr11) [reply docs/package_reference/sparse_encoder/evaluation.md:38]: patch with what's say above

File: docs/package_reference/sparse_encoder/losses.md [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+## SparseMarginMSELoss
+```{eval-rst}
+.. autoclass:: sentence_transformers.sparse_encoder.losses.SparseMarginMSELoss
+   :members:
+```
+
+## SpladeLoss
+```{eval-rst}
+.. autoclass:: sentence_transformers.sparse_encoder.losses.SpladeLoss
+   :members:
```
[the remark this answers is not in the document]
Comment (arthurbr11) [reply docs/package_reference/sparse_encoder/losses.md:41]: Yes just did like it's done in ST so i get rid of the members mention in the .rst

File: docs/package_reference/sparse_encoder/sampler.md
[the remark this answers is not in the document]
Comment (arthurbr11) [reply docs/package_reference/sparse_encoder/sampler.md:1]: No it is indeed more clear just didn't know how to do that :sweat_smile:

File: sentence_transformers/sparse_encoder/SparseEncoder.py (hunk 5 above)
Comment (arthurbr11) [reply sentence_transformers/sparse_encoder/SparseEncoder.py:260]: Hmm i think for now as we only support torch COO it's fair enough to not precise

File: sentence_transformers/sparse_encoder/SparseEncoder.py (hunk 6 above)
Comment (arthurbr11) [reply sentence_transformers/sparse_encoder/SparseEncoder.py:262]: i like `save_to_cpu`

File: sentence_transformers/sparse_encoder/callbacks/splade_callbacks.py (hunk 8 above)
Comment (arthurbr11) [reply sentence_transformers/sparse_encoder/callbacks/splade_callbacks.py:33]: I agree doing it rn !

File: sentence_transformers/sparse_encoder/losses/CSRReconstructionLoss.py [hunk 16]
```diff
@@ -20,15 +20,36 @@ def normalized_mean_squared_error(reconstruction: torch.Tensor, original_input:
...
+    Requirements:
+        1. The model must be configured to output the necessary reconstruction components
+        2. Used with SparseEncoder models that implement compositional sparse autoencoding
+
+    Relations:
+        - Used as a component within :class:`CSRLoss` combined with a contrastive loss
+
+    Example:
+        ::
+            This loss is typically used within the :class:`CSRLoss` class, which combines it with other loss components.
```
[the remark this answers is not in the document]
Comment (arthurbr11) [reply sentence_transformers/sparse_encoder/losses/CSRReconstructionLoss.py]: I think you don't need it if it's import in the file

File: docs/package_reference/sparse_encoder/sampler.md
[the remark this answers is not in the document]
Comment (arthurbr11) [reply docs/package_reference/sparse_encoder/sampler.md:1]: Just found how to do it, Done!

File: docs/package_reference/sparse_encoder/SparseEncoder.md [hunk 17]
```diff
@@ -0,0 +1,21 @@
+# SparseEncoder
+
+## SparseEncoder
+```{eval-rst}
+.. autoclass:: sentence_transformers.sparse_encoder.SparseEncoder
+   :members:
```
[the remark this answers is not in the document]
Comment (arthurbr11) [reply docs/package_reference/sparse_encoder/SparseEncoder.md:6]: So do you think we should exclude all the members that aren't overrides ?

**Comment by tomaarsen** [comment]
I believe
```python
    @max_seq_length.setter
    def max_seq_length(self, value) -> None:
        """
        Property to set the maximal input sequence length for the model. Longer inputs will be truncated.
        """
        self._first_module().max_seq_length = value
```
is necessary now that `max_seq_length` is overridden? I get `AttributeError: property 'max_seq_length' of 'SparseEncoder' object has no setter` if I try to update it after loading.

**Comment by arthurbr11** [comment]
> Also as we discussed, a `model.decode` should be valuable, accepting e.g. 1 `embeddings` or perhaps 2 sets of embeddings (queries, docs) that get multiplied for a joint embedding? Maybe then we need to call something akin to `model.similarity` except instead of 1 similarity score we need a joint embedding somehow.

I agree that a somehow `pointwise` similarity can be really valuable to have the importance of each token in the scores.

File: sentence_transformers/sparse_encoder/losses/RegularizerLoss.py [hunk 18]
```diff
@@ -41,13 +43,20 @@ def forward(self, sentence_features: Iterable[dict[str, Tensor]], labels: Tensor
         return self.compute_loss_from_embeddings(embeddings)
 
     def compute_loss_from_embeddings(self, embeddings: list[torch.Tensor], embeddings_type: str) -> torch.Tensor:
-        anchors = embeddings[0]  # (batch_size, embedding_dim)
-        candidates = torch.cat(embeddings[1:])  # (batch_size * (1 + num_negatives), embedding_dim)
+        if embeddings_type == "query":
+            embeddings_to_use = embeddings[0]  # (batch_size, embedding_dim)
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/losses/RegularizerLoss.py:47]: Nitpick: I think you can just do `embeddings` instead of `embeddings_to_use`

File: sentence_transformers/sparse_encoder/losses/RegularizerLoss.py
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/losses/RegularizerLoss.py:1]: We might have to rename this file back to just FlopsLoss then?

**Comment by tomaarsen** [comment]
I think we can remove `get_top_k_token_weight` from `util.py` now.

File: sentence_transformers/sparse_encoder/SparseEncoder.py [hunk 19]
```diff
@@ -564,3 +969,108 @@ def get_sparsity_stats(embeddings: torch.Tensor) -> dict[str, float]:
...
+    def intersection(
+        self,
+        embeddings_1: torch.Tensor,
+        embeddings_2: torch.Tensor,
+    ):
+        """
+        Compute the intersection of two sparse embeddings.
+
+        Args:
+            embeddings_1 (torch.Tensor): First embedding tensor (batch_1, vocab).
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/SparseEncoder.py:1008]:
```suggestion
            embeddings_1 (torch.Tensor): First embedding tensor (vocab).
```
The first embedding must be 1-dimensional

**Comment by tomaarsen** [comment]
Long explanation incoming, but you can update
```python
@pytest.mark.slow
def test_other_models_slow() -> None:
    pretrained_model_score_slow("average_word_embeddings_komninos", 61.56)
```
to
```python
@pytest.mark.slow
def test_other_models_slow() -> None:
    pretrained_model_score_slow("average_word_embeddings_komninos", 60.98)
```
in `tests/test_pretrained_stsb.py`.

Your commit 5d86a5476067ec414fe4a9a8495060a7572166d4 updated how the distances/similarities are computed from sklearn to something in `util.py`. It looks like there's 1 edge case that was updated, which is causing test failures with the `slow` mark: if one of the embeddings is just a 0-vector.
The previous `1 - paired_cosine_distance(...)` would result in a distance of 0 a.k.a. a similarity of 1, whereas the new `paired_cos_sim` results in a similarity of 0. This means that the overall scores update in this very niche case, but I think the new situation is actually preferable: I want a 0-vector to have a similarity of 0 to all other embeddings, instead of a similarity of 1. 

Furthermore, it only happens with 0-vectors, which don't really occur, so I have no problem with just updating the expected score and calling it a day.

**Comment by tomaarsen** [comment]
I see that I didn't seem to commit one of my changes to the css:
```css
.components > .box > .header {
    border-style: solid;
    border-width: 1px;
    border-radius: 0.5rem;
    border-color: rgb(55 65 81);
    padding: 0.5rem 0.2rem;
    text-align: center;
    margin-bottom: 0.5rem;
    font-weight: bold;
    color: white;
}
```
(the padding `0.2rem` is new)
Could you try and include this in `docs/_static/css/custom.css`? It allows for the Training Overview to be on 1 line on my Firefox & Edge: 
image

File: docs/package_reference/sparse_encoder/models.md [hunk 20]
```diff
@@ -0,0 +1,24 @@
+# Models
```
Comment (tomaarsen) [inline docs/package_reference/sparse_encoder/models.md:1]:
```suggestion
# Modules
```
Same as above

File: docs/package_reference/sparse_encoder/models.md [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+
+`sentence_transformers.sparse_encoder.models` defines different building blocks, that can be used to create SparseEncoder networks from scratch. For more details, see [Training Overview](../../sparse_encoder/training_overview.md).
+Note that models from `sentence_transformers.models` can be used to such as `sentence_transformers.models.Transformer` see [SentenceTransformer Models](../sentence_transformer/models.md)
```
Comment (tomaarsen) [inline docs/package_reference/sparse_encoder/models.md:4]:
```suggestion
Note that modules from `sentence_transformers.models` can also be used for Sparse models, such as `sentence_transformers.models.Transformer` from [SentenceTransformer > Modules](../sentence_transformer/models.md)
```

**Review by tomaarsen — COMMENTED** [review]
Forgot to submit this review, oops

File: sentence_transformers/sparse_encoder/models/IDF.py [hunk 22]
```diff
@@ -1,31 +1,62 @@
...
+    """
+
+    config_keys: list[str] = ["frozen"]
 
-class IDF(torch.nn.Module):
     def __init__(
         self,
-        weight: torch.Tensor,
         tokenizer: PreTrainedTokenizer,
+        weight: torch.Tensor | None,
```
[the remark this answers is not in the document]
Comment (tomaarsen) [reply sentence_transformers/sparse_encoder/models/IDF.py]: Already resolved

File: examples/sparse_encoder/training/train_nq.py [hunk 23]
```diff
@@ -44,7 +40,7 @@ def main():
     logging.info(eval_dataset)
 
     # 3. Initialize the loss
-    loss = CSRLoss(
+    loss = losses.CSRLoss(
         model=model,
         beta=0.1,  # Weight for auxiliary loss
         gamma=1,  # Weight for ranking loss
```
Comment (tomaarsen) [inline examples/sparse_encoder/training/train_nq.py:46]: Not sure if this file is ready, but there's no `loss` passed, and `scale` doesn't exist here anymore

Comment (arthurbr11) [reply examples/sparse_encoder/training/train_nq.py:46]: I haven't update those yet no. I only updated the ones in a sub-folder under training

**Comment by tomaarsen** [comment]
Don't be afraid to just merge this one into `sparse_implementation` and e.g. do the `tests` work in a new branch again.
